### ISSUE TITLE
vim-go: offset X is past the end of the file Y

### ISSUE THREAD

**Issue by shayan-7** [fatih/vim-go#3206]
When you copy a snippet code from a website styled by CSS, and paste in VIM and save the file, the vim-go shows the following error at the bottom of the code:

```offset 2820 is past the end of the file 2790```

![image](https://user-images.githubusercontent.com/39003982/115524773-b5e53b00-a2a3-11eb-81db-b9caaaf517d9.png)

(As you see in this pic, I'm using other plugins but I also have tried without any other plugins but vim-go)

### What did you do?

I copied and pasted a snippet code from https://gowebexamples.com/templates.
You can try any styled codes.

####  `vimrc` you used to reproduce:

The only code of vimrc could be:

```vim
call plug#begin('~/.vim/plugged')
Plug 'fatih/vim-go'
call plug#end()
```
</details>

#### Go version (`go version`):
go version go1.16.2 linux/amd64

#### Go environment
```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/shayan/.cache/go-build"
GOENV="/home/shayan/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/shayan/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/shayan/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.2"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3164144905=/tmp/go-build -gno-record-gcc-switches"
```

#### gopls version
```
golang.org/x/tools/gopls v0.6.10
    golang.org/x/tools/gopls@v0.6.10 h1:8Ebz8PymS2umcuCFhoz67unyJfWsipjTIrkBUF9kypg=
```

**Comment by bhcleek** [comment]
Can you supply the gopls logs by enabling them by running `let g:go_debug=['lsp']` before you copy and paste?

**Comment by shayan-7** [comment]
I did and the log was:

```
===== sent =====
Content-Length: 927

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\n\nimport (\n    \"html/template\"\n    \"net/http\"\n)\n\n\n\ntype Todo struct {\n    Title string\n    Done  bool\n}\n\n\n\ntype TodoPageData struct {\n    PageTitle string\n    Todos     []Todo\n}\n\n\n\nfunc main() {\n    tmpl := template.Must(template.ParseFiles(\"layout.html\"))\n    http.HandleFunc(\"/\", func(w http.ResponseWriter, r *http.Request) {\n        data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/goprac/concurrency/main.go","version":1439}}}
===== sent =====
Content-Length: 185

{"method":"textDocument/formatting","jsonrpc":"2.0","id":8,"params":{"options":{"insertSpaces":false},"textDocument":{"uri":"file:///home/shayan/workspace/goprac/concurrency/main.go"}}}
===== received =====
Content-Length: 3046

{"jsonrpc":"2.0","result":[{"range":{"start":{"line":2,"character":0},"end":{"line":2,"character":0}},"newText":"import (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\""},{"range":{"start":{"line":5,"character":5},"end":{"line":5,"character":5}},"newText":"html/template\"\n    \"html/template\"\n    \""},{"range":{"start":{"line":6,"character":0},"end":{"line":6,"character":0}},"newText":"    \"html/template\"\n"},{"range":{"start":{"line":7,"character":0},"end":{"line":7,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":9,"character":0},"end":{"line":9,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":11,"character":4},"end":{"line":11,"character":4}},"newText":"\"html/template\"\n    "},{"range":{"start":{"line":11,"character":16},"end":{"line":11,"character":16}},"newText":"\n    \"html/template\""},{"range":{"start":{"line":14,"character":0},"end":{"line":14,"character":0}},"newText":""},{"range":{"start":{"line":14,"character":0},"end":{"line":14,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":16,"character":0},"end":{"line":17,"character":0}},"newText":""},{"range":{"start":{"line":17,"character":0},"end":{"line":17,"character":0}},"newText":"func main() {\n    \"net/http\""},{"range":{"start":{"line":17,"character":1},"end":{"line":17,"character":1}},"newText":"    \"net/http\""},{"range":{"start":{"line":17,"character":3},"end":{"line":17,"character":3}},"newText":"    \"net/http\""},{"range":{"start":{"line":17,"character":13},"end":{"line":17,"character":18}},"newText":""},{"range":{"start":{"line":17,"character":18},"end":{"line":17,"character":18}},"newText":"    \"net/http\""},{"range":{"start":{"line":18,"character":12},"end":{"line":18,"character":12}},"newText":"    \"net/http\""},{"range":{"start":{"line":19,"character":7},"end":{"line":19,"character":7}},"newText":"    \"net/http\""},{"range":{"start":{"line":19,"character":10},"end":{"line":19,"character":10}},"newText":"    \"net/http\""},{"range":{"start":{"line":24,"character":4},"end":{"line":24,"character":4}},"newText":""},{"range":{"start":{"line":24,"character":4},"end":{"line":24,"character":4}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":8},"end":{"line":25,"character":8}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":24},"end":{"line":25,"character":24}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":27},"end":{"line":25,"character":35}},"newText":""},{"range":{"start":{"line":25,"character":35},"end":{"line":25,"character":35}},"newText":")"},{"range":{"start":{"line":25,"character":36},"end":{"line":26,"character":37}},"newText":""},{"range":{"start":{"line":26,"character":39},"end":{"line":26,"character":47}},"newText":""},{"range":{"start":{"line":26,"character":50},"end":{"line":26,"character":50}},"newText":""},{"range":{"start":{"line":26,"character":59},"end":{"line":27,"character":6}},"newText":""}],"id":8}
===== sent =====
Content-Length: 1266

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\nimport (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\"\nimport (\n    \"html/template\"\n    \"html/template\"\n    \"html/template\"\n    \"net/http\"\n    \"html/template\"\n)\n    \"html/template\"\n\n    \"html/template\"\ntype Todo struct {\n    \"html/template\"\n    Title string\n    \"html/template\"\n    Done  bool\n}\n    \"html/template\"\n\nfunc main() {\n    \"net/http\"t    \"net/http\"yp    \"net/http\"e TodoPage    \"net/http\"struct {\n    PageTitl    \"net/http\"e string\n    Tod    \"net/http\"os     \"net/http\"    []Todo\n}\n\n\n\nfunc    \"net/http\" main() {\n    tmpl    \"net/http\" := template.Mus    \"net/http\"t(t)PReiter, r *htt  data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/goprac/concurrency/main.go","version":1506}}}
===== sent =====
Content-Length: 273

{"method":"textDocument/codeAction","jsonrpc":"2.0","id":9,"params":{"context":{"only":["source.organizeImports"]},"range":{"end":{"character":0,"line":48},"start":{"character":0,"line":0}},"textDocument":{"uri":"file:///home/shayan/workspace/goprac/concurrency/main.go"}}}
===== received =====
Content-Length: 315

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":1,"message":"2021/04/22 20:14:00 imports fixes: AllImportsFixes: /home/shayan/workspace/goprac/concurrency/main.go:7:5: expected declaration, found \"html/template\" (and 2 more errors)\n\tfile=\"/home/shayan/workspace/goprac/concurrency/main.go\"\n"}}
===== received =====
Content-Length: 255

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2021/04/22 20:14:00 go/packages.Load\n\tsnapshot=12\n\tdirectory=/home/shayan/workspace/goprac\n\tquery=[file=/home/shayan/workspace/goprac/concurrency/main.go]\n\tpackages=1\n"}}
===== received =====
Content-Length: 247

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2021/04/22 20:14:00 go/packages.Load\n\tsnapshot=12\n\tpackage=\"github.com/shayan-7/goprac/concurrency\"\n\tfiles=[/home/shayan/workspace/goprac/concurrency/main.go]\n"}}
===== received =====
Content-Length: 332

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/shayan/workspace/goprac/concurrency/main.go","version":1506,"diagnostics":[{"range":{"start":{"line":6,"character":4},"end":{"line":6,"character":4}},"severity":1,"source":"syntax","message":"expected declaration, found \"html/template\""}]}}
===== received =====
Content-Length: 38

{"jsonrpc":"2.0","result":null,"id":9}
===== received =====
Content-Length: 255

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2021/04/22 20:14:01 go/packages.Load\n\tsnapshot=12\n\tdirectory=/home/shayan/workspace/goprac\n\tquery=[file=/home/shayan/workspace/goprac/concurrency/main.go]\n\tpackages=1\n"}}
===== received =====
Content-Length: 247

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2021/04/22 20:14:01 go/packages.Load\n\tsnapshot=12\n\tpackage=\"github.com/shayan-7/goprac/concurrency\"\n\tfiles=[/home/shayan/workspace/goprac/concurrency/main.go]\n"}}
===== sent =====
Content-Length: 1266

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\nimport (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\"\nimport (\n    \"html/template\"\n    \"html/template\"\n    \"html/template\"\n    \"net/http\"\n    \"html/template\"\n)\n    \"html/template\"\n\n    \"html/template\"\ntype Todo struct {\n    \"html/template\"\n    Title string\n    \"html/template\"\n    Done  bool\n}\n    \"html/template\"\n\nfunc main() {\n    \"net/http\"t    \"net/http\"yp    \"net/http\"e TodoPage    \"net/http\"struct {\n    PageTitl    \"net/http\"e string\n    Tod    \"net/http\"os     \"net/http\"    []Todo\n}\n\n\n\nfunc    \"net/http\" main() {\n    tmpl    \"net/http\" := template.Mus    \"net/http\"t(t)PReiter, r *htt  data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/goprac/concurrency/main.go","version":1507}}}
===== received =====
Content-Length: 462

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":1,"message":"2021/04/22 20:14:01 diagnosing go.mod: err: exit status 1: stderr: go: downloading gopkg.in/check.v1 v1.0.0-20190902080502-41f04d3bba15\ngithub.com/shayan-7/goprac/file imports\n\tgithub.com/gin-gonic/gin imports\n\tgithub.com/gin-gonic/gin/binding imports\n\tgopkg.in/yaml.v2 tested by\n\tgopkg.in/yaml.v2.test imports\n\tgopkg.in/check.v1: module lookup disabled by GOPROXY=off\n\n"}}
```

**Comment by bhcleek** [comment]
I've tried duplicating this several times without success.

**Comment by shayan-7** [comment]
Have you tried copying and pasting from the exact page I linked in the issue?
Is there any other way to figure out what the problem is?

**Comment by bhcleek** [comment]
Yes, I copied and pasted from the page you linked in the issue.

It's possible that you're using an older version of vim-go; since you didn't report which revision you're using, it's hard to know for sure, though.

Can you try disabling `g:go_imports_autosave` and try again? It would help if we could narrow down which operation this is coming from.

Also, from the logs, it looks like what's being sent to gopls isn't the same as what's on the page that you linked. The `textDocument/didChange` requests look like what's in your buffer is
```
package main

import (
	"html/template"
	"net/http"
)
    "html/template"

    "html/template"
import (
    "html/template"
    "html/template"
    "html/template"
    "net/http"
    "html/template"
)
    "html/template"

    "html/template"
type Todo struct {
    "html/template"
    Title string
    "html/template"
    Done  bool
}
    "html/template"

func main() {
    "net/http"t    "net/http"yp    "net/http"e TodoPage    "net/http"struct {
    PageTitl    "net/http"e string
    Tod    "net/http"os     "net/http"    []Todo
}



func    "net/http" main() {
    tmpl    "net/http" := template.Mus    "net/http"t(t)PReiter, r *htt  data := TodoPageData{
            PageTitle: "My TODO list",
            Todos: []Todo{
                {Title: "Task 1", Done: false},
                {Title: "Task 2", Done: true},
                {Title: "Task 3", Done: true},
            },
        }
        tmpl.Execute(w, data)
    })
    http.ListenAndServe(":80", nil)
}

```

This is also supported by the diagnostic messages reported by `gopls`:
```
{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":1,"message":"2021/04/22 20:14:00 imports fixes: AllImportsFixes: /home/shayan/workspace/goprac/concurrency/main.go:7:5: expected declaration, found \"html/template\" (and 2 more errors)\n\tfile=\"/home/shayan/workspace/goprac/concurrency/main.go\"\n"}}
```

If I had to guess, I'd suspect that maybe you have something in the file that you start with, that you then use blockwise visual mode to select some text, and that you're then attempting to paste.

Can you tell me about the precise steps you're using to duplicate the behavior you reported?

**Comment by shayan-7** [comment]
First I have to thank you for your accurate and elegant reply.
Second That's exactly what sometimes happens from I copy a snippet code ( I mean the messy code after `:wq` ) .
Would you tell me how I should check the version of `vim-go`?
I tried to check by `git describe --tags` in the plugin directory of `vim-go` ( ~/.vim/plugged/vim-go ) .

**Comment by bhcleek** [comment]
`git rev-parse HEAD` in the plugin directory of vim-go should tell you what revision you're on.

How exactly are you pasting? Precise steps would be helpful to understand what's happening. Do you see the messed up text immediately after pasting or only after save?

**Comment by shayan-7** [comment]
The `git rev-parse HEAD` returns `885243576bba991cee4b88b212351aead9044bdc`.
I select the text by dragging the mouse pointer and copy by `Ctrl+c` and paste by `Ctrl+shift+v` in Vim.
The messed-up text appears just after saving the file.

**Comment by shayan-7** [comment]
The log once I paste the code in vim:

```
===== sent =====
Content-Length: 918

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\n\nimport (\n    \"html/template\"\n    \"net/http\"\n)\n\n\n\ntype Todo struct {\n    Title string\n    Done  bool\n}\n\n\n\ntype TodoPageData struct {\n    PageTitle string\n    Todos     []Todo\n}\n\n\n\nfunc main() {\n    tmpl := template.Must(template.ParseFiles(\"layout.html\"))\n    http.HandleFunc(\"/\", func(w http.ResponseWriter, r *http.Request) {\n        data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1371}}}
===== received =====
Content-Length: 248

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2021/04/26 15:18:19 go/packages.Load\n\tsnapshot=7\n\tdirectory=/home/shayan/workspace/vimgotest\n\tquery=[file=/home/shayan/workspace/vimgotest/main.go]\n\tpackages=1\n"}}
===== received =====
Content-Length: 221

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2021/04/26 15:18:19 go/packages.Load\n\tsnapshot=7\n\tpackage=\"command-line-arguments\"\n\tfiles=[/home/shayan/workspace/vimgotest/main.go]\n"}}
===== received =====
Content-Length: 159

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1371,"diagnostics":[]}}
===== sent =====
Content-Length: 176

{"method":"textDocument/formatting","jsonrpc":"2.0","id":3,"params":{"options":{"insertSpaces":false},"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go"}}}
===== received =====
Content-Length: 3046

{"jsonrpc":"2.0","result":[{"range":{"start":{"line":2,"character":0},"end":{"line":2,"character":0}},"newText":"import (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\""},{"range":{"start":{"line":5,"character":5},"end":{"line":5,"character":5}},"newText":"html/template\"\n    \"html/template\"\n    \""},{"range":{"start":{"line":6,"character":0},"end":{"line":6,"character":0}},"newText":"    \"html/template\"\n"},{"range":{"start":{"line":7,"character":0},"end":{"line":7,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":9,"character":0},"end":{"line":9,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":11,"character":4},"end":{"line":11,"character":4}},"newText":"\"html/template\"\n    "},{"range":{"start":{"line":11,"character":16},"end":{"line":11,"character":16}},"newText":"\n    \"html/template\""},{"range":{"start":{"line":14,"character":0},"end":{"line":14,"character":0}},"newText":""},{"range":{"start":{"line":14,"character":0},"end":{"line":14,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":16,"character":0},"end":{"line":17,"character":0}},"newText":""},{"range":{"start":{"line":17,"character":0},"end":{"line":17,"character":0}},"newText":"func main() {\n    \"net/http\""},{"range":{"start":{"line":17,"character":1},"end":{"line":17,"character":1}},"newText":"    \"net/http\""},{"range":{"start":{"line":17,"character":3},"end":{"line":17,"character":3}},"newText":"    \"net/http\""},{"range":{"start":{"line":17,"character":13},"end":{"line":17,"character":18}},"newText":""},{"range":{"start":{"line":17,"character":18},"end":{"line":17,"character":18}},"newText":"    \"net/http\""},{"range":{"start":{"line":18,"character":12},"end":{"line":18,"character":12}},"newText":"    \"net/http\""},{"range":{"start":{"line":19,"character":7},"end":{"line":19,"character":7}},"newText":"    \"net/http\""},{"range":{"start":{"line":19,"character":10},"end":{"line":19,"character":10}},"newText":"    \"net/http\""},{"range":{"start":{"line":24,"character":4},"end":{"line":24,"character":4}},"newText":""},{"range":{"start":{"line":24,"character":4},"end":{"line":24,"character":4}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":8},"end":{"line":25,"character":8}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":24},"end":{"line":25,"character":24}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":27},"end":{"line":25,"character":35}},"newText":""},{"range":{"start":{"line":25,"character":35},"end":{"line":25,"character":35}},"newText":")"},{"range":{"start":{"line":25,"character":36},"end":{"line":26,"character":37}},"newText":""},{"range":{"start":{"line":26,"character":39},"end":{"line":26,"character":47}},"newText":""},{"range":{"start":{"line":26,"character":50},"end":{"line":26,"character":50}},"newText":""},{"range":{"start":{"line":26,"character":59},"end":{"line":27,"character":6}},"newText":""}],"id":3}
===== sent =====
Content-Length: 1257

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\nimport (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\"\nimport (\n    \"html/template\"\n    \"html/template\"\n    \"html/template\"\n    \"net/http\"\n    \"html/template\"\n)\n    \"html/template\"\n\n    \"html/template\"\ntype Todo struct {\n    \"html/template\"\n    Title string\n    \"html/template\"\n    Done  bool\n}\n    \"html/template\"\n\nfunc main() {\n    \"net/http\"t    \"net/http\"yp    \"net/http\"e TodoPage    \"net/http\"struct {\n    PageTitl    \"net/http\"e string\n    Tod    \"net/http\"os     \"net/http\"    []Todo\n}\n\n\n\nfunc    \"net/http\" main() {\n    tmpl    \"net/http\" := template.Mus    \"net/http\"t(t)PReiter, r *htt  data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1438}}}
===== received =====
Content-Length: 323

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1438,"diagnostics":[{"range":{"start":{"line":6,"character":4},"end":{"line":6,"character":4}},"severity":1,"source":"syntax","message":"expected declaration, found \"html/template\""}]}}
===== sent =====
Content-Length: 1257

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\nimport (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\"\nimport (\n    \"html/template\"\n    \"html/template\"\n    \"html/template\"\n    \"net/http\"\n    \"html/template\"\n)\n    \"html/template\"\n\n    \"html/template\"\ntype Todo struct {\n    \"html/template\"\n    Title string\n    \"html/template\"\n    Done  bool\n}\n    \"html/template\"\n\nfunc main() {\n    \"net/http\"t    \"net/http\"yp    \"net/http\"e TodoPage    \"net/http\"struct {\n    PageTitl    \"net/http\"e string\n    Tod    \"net/http\"os     \"net/http\"    []Todo\n}\n\n\n\nfunc    \"net/http\" main() {\n    tmpl    \"net/http\" := template.Mus    \"net/http\"t(t)PReiter, r *htt  data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1439}}}
===== sent =====
Content-Length: 918

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\n\nimport (\n    \"html/template\"\n    \"net/http\"\n)\n\n\n\ntype Todo struct {\n    Title string\n    Done  bool\n}\n\n\n\ntype TodoPageData struct {\n    PageTitle string\n    Todos     []Todo\n}\n\n\n\nfunc main() {\n    tmpl := template.Must(template.ParseFiles(\"layout.html\"))\n    http.HandleFunc(\"/\", func(w http.ResponseWriter, r *http.Request) {\n        data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1507}}}
===== received =====
Content-Length: 159

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1507,"diagnostics":[]}}
```

The log that is generated on saving the file:

```
===== sent =====
Content-Length: 918

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\n\nimport (\n    \"html/template\"\n    \"net/http\"\n)\n\n\n\ntype Todo struct {\n    Title string\n    Done  bool\n}\n\n\n\ntype TodoPageData struct {\n    PageTitle string\n    Todos     []Todo\n}\n\n\n\nfunc main() {\n    tmpl := template.Must(template.ParseFiles(\"layout.html\"))\n    http.HandleFunc(\"/\", func(w http.ResponseWriter, r *http.Request) {\n        data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1371}}}
===== received =====
Content-Length: 248

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2021/04/26 15:18:19 go/packages.Load\n\tsnapshot=7\n\tdirectory=/home/shayan/workspace/vimgotest\n\tquery=[file=/home/shayan/workspace/vimgotest/main.go]\n\tpackages=1\n"}}
===== received =====
Content-Length: 221

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2021/04/26 15:18:19 go/packages.Load\n\tsnapshot=7\n\tpackage=\"command-line-arguments\"\n\tfiles=[/home/shayan/workspace/vimgotest/main.go]\n"}}
===== received =====
Content-Length: 159

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1371,"diagnostics":[]}}
===== sent =====
Content-Length: 176

{"method":"textDocument/formatting","jsonrpc":"2.0","id":3,"params":{"options":{"insertSpaces":false},"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go"}}}
===== received =====
Content-Length: 3046

{"jsonrpc":"2.0","result":[{"range":{"start":{"line":2,"character":0},"end":{"line":2,"character":0}},"newText":"import (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\""},{"range":{"start":{"line":5,"character":5},"end":{"line":5,"character":5}},"newText":"html/template\"\n    \"html/template\"\n    \""},{"range":{"start":{"line":6,"character":0},"end":{"line":6,"character":0}},"newText":"    \"html/template\"\n"},{"range":{"start":{"line":7,"character":0},"end":{"line":7,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":9,"character":0},"end":{"line":9,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":11,"character":4},"end":{"line":11,"character":4}},"newText":"\"html/template\"\n    "},{"range":{"start":{"line":11,"character":16},"end":{"line":11,"character":16}},"newText":"\n    \"html/template\""},{"range":{"start":{"line":14,"character":0},"end":{"line":14,"character":0}},"newText":""},{"range":{"start":{"line":14,"character":0},"end":{"line":14,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":16,"character":0},"end":{"line":17,"character":0}},"newText":""},{"range":{"start":{"line":17,"character":0},"end":{"line":17,"character":0}},"newText":"func main() {\n    \"net/http\""},{"range":{"start":{"line":17,"character":1},"end":{"line":17,"character":1}},"newText":"    \"net/http\""},{"range":{"start":{"line":17,"character":3},"end":{"line":17,"character":3}},"newText":"    \"net/http\""},{"range":{"start":{"line":17,"character":13},"end":{"line":17,"character":18}},"newText":""},{"range":{"start":{"line":17,"character":18},"end":{"line":17,"character":18}},"newText":"    \"net/http\""},{"range":{"start":{"line":18,"character":12},"end":{"line":18,"character":12}},"newText":"    \"net/http\""},{"range":{"start":{"line":19,"character":7},"end":{"line":19,"character":7}},"newText":"    \"net/http\""},{"range":{"start":{"line":19,"character":10},"end":{"line":19,"character":10}},"newText":"    \"net/http\""},{"range":{"start":{"line":24,"character":4},"end":{"line":24,"character":4}},"newText":""},{"range":{"start":{"line":24,"character":4},"end":{"line":24,"character":4}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":8},"end":{"line":25,"character":8}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":24},"end":{"line":25,"character":24}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":27},"end":{"line":25,"character":35}},"newText":""},{"range":{"start":{"line":25,"character":35},"end":{"line":25,"character":35}},"newText":")"},{"range":{"start":{"line":25,"character":36},"end":{"line":26,"character":37}},"newText":""},{"range":{"start":{"line":26,"character":39},"end":{"line":26,"character":47}},"newText":""},{"range":{"start":{"line":26,"character":50},"end":{"line":26,"character":50}},"newText":""},{"range":{"start":{"line":26,"character":59},"end":{"line":27,"character":6}},"newText":""}],"id":3}
===== sent =====
Content-Length: 1257

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\nimport (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\"\nimport (\n    \"html/template\"\n    \"html/template\"\n    \"html/template\"\n    \"net/http\"\n    \"html/template\"\n)\n    \"html/template\"\n\n    \"html/template\"\ntype Todo struct {\n    \"html/template\"\n    Title string\n    \"html/template\"\n    Done  bool\n}\n    \"html/template\"\n\nfunc main() {\n    \"net/http\"t    \"net/http\"yp    \"net/http\"e TodoPage    \"net/http\"struct {\n    PageTitl    \"net/http\"e string\n    Tod    \"net/http\"os     \"net/http\"    []Todo\n}\n\n\n\nfunc    \"net/http\" main() {\n    tmpl    \"net/http\" := template.Mus    \"net/http\"t(t)PReiter, r *htt  data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1438}}}
===== received =====
Content-Length: 323

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1438,"diagnostics":[{"range":{"start":{"line":6,"character":4},"end":{"line":6,"character":4}},"severity":1,"source":"syntax","message":"expected declaration, found \"html/template\""}]}}
===== sent =====
Content-Length: 1257

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\nimport (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\"\nimport (\n    \"html/template\"\n    \"html/template\"\n    \"html/template\"\n    \"net/http\"\n    \"html/template\"\n)\n    \"html/template\"\n\n    \"html/template\"\ntype Todo struct {\n    \"html/template\"\n    Title string\n    \"html/template\"\n    Done  bool\n}\n    \"html/template\"\n\nfunc main() {\n    \"net/http\"t    \"net/http\"yp    \"net/http\"e TodoPage    \"net/http\"struct {\n    PageTitl    \"net/http\"e string\n    Tod    \"net/http\"os     \"net/http\"    []Todo\n}\n\n\n\nfunc    \"net/http\" main() {\n    tmpl    \"net/http\" := template.Mus    \"net/http\"t(t)PReiter, r *htt  data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1439}}}
===== sent =====
Content-Length: 918

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\n\nimport (\n    \"html/template\"\n    \"net/http\"\n)\n\n\n\ntype Todo struct {\n    Title string\n    Done  bool\n}\n\n\n\ntype TodoPageData struct {\n    PageTitle string\n    Todos     []Todo\n}\n\n\n\nfunc main() {\n    tmpl := template.Must(template.ParseFiles(\"layout.html\"))\n    http.HandleFunc(\"/\", func(w http.ResponseWriter, r *http.Request) {\n        data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1507}}}
===== received =====
Content-Length: 159

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1507,"diagnostics":[]}}
===== sent =====
Content-Length: 176

{"method":"textDocument/formatting","jsonrpc":"2.0","id":4,"params":{"options":{"insertSpaces":false},"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go"}}}
===== received =====
Content-Length: 3046

{"jsonrpc":"2.0","result":[{"range":{"start":{"line":2,"character":0},"end":{"line":2,"character":0}},"newText":"import (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\""},{"range":{"start":{"line":5,"character":5},"end":{"line":5,"character":5}},"newText":"html/template\"\n    \"html/template\"\n    \""},{"range":{"start":{"line":6,"character":0},"end":{"line":6,"character":0}},"newText":"    \"html/template\"\n"},{"range":{"start":{"line":7,"character":0},"end":{"line":7,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":9,"character":0},"end":{"line":9,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":11,"character":4},"end":{"line":11,"character":4}},"newText":"\"html/template\"\n    "},{"range":{"start":{"line":11,"character":16},"end":{"line":11,"character":16}},"newText":"\n    \"html/template\""},{"range":{"start":{"line":14,"character":0},"end":{"line":14,"character":0}},"newText":""},{"range":{"start":{"line":14,"character":0},"end":{"line":14,"character":0}},"newText":"    \"html/template\""},{"range":{"start":{"line":16,"character":0},"end":{"line":17,"character":0}},"newText":""},{"range":{"start":{"line":17,"character":0},"end":{"line":17,"character":0}},"newText":"func main() {\n    \"net/http\""},{"range":{"start":{"line":17,"character":1},"end":{"line":17,"character":1}},"newText":"    \"net/http\""},{"range":{"start":{"line":17,"character":3},"end":{"line":17,"character":3}},"newText":"    \"net/http\""},{"range":{"start":{"line":17,"character":13},"end":{"line":17,"character":18}},"newText":""},{"range":{"start":{"line":17,"character":18},"end":{"line":17,"character":18}},"newText":"    \"net/http\""},{"range":{"start":{"line":18,"character":12},"end":{"line":18,"character":12}},"newText":"    \"net/http\""},{"range":{"start":{"line":19,"character":7},"end":{"line":19,"character":7}},"newText":"    \"net/http\""},{"range":{"start":{"line":19,"character":10},"end":{"line":19,"character":10}},"newText":"    \"net/http\""},{"range":{"start":{"line":24,"character":4},"end":{"line":24,"character":4}},"newText":""},{"range":{"start":{"line":24,"character":4},"end":{"line":24,"character":4}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":8},"end":{"line":25,"character":8}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":24},"end":{"line":25,"character":24}},"newText":"    \"net/http\""},{"range":{"start":{"line":25,"character":27},"end":{"line":25,"character":35}},"newText":""},{"range":{"start":{"line":25,"character":35},"end":{"line":25,"character":35}},"newText":")"},{"range":{"start":{"line":25,"character":36},"end":{"line":26,"character":37}},"newText":""},{"range":{"start":{"line":26,"character":39},"end":{"line":26,"character":47}},"newText":""},{"range":{"start":{"line":26,"character":50},"end":{"line":26,"character":50}},"newText":""},{"range":{"start":{"line":26,"character":59},"end":{"line":27,"character":6}},"newText":""}],"id":4}
===== sent =====
Content-Length: 1257

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\nimport (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\"\nimport (\n    \"html/template\"\n    \"html/template\"\n    \"html/template\"\n    \"net/http\"\n    \"html/template\"\n)\n    \"html/template\"\n\n    \"html/template\"\ntype Todo struct {\n    \"html/template\"\n    Title string\n    \"html/template\"\n    Done  bool\n}\n    \"html/template\"\n\nfunc main() {\n    \"net/http\"t    \"net/http\"yp    \"net/http\"e TodoPage    \"net/http\"struct {\n    PageTitl    \"net/http\"e string\n    Tod    \"net/http\"os     \"net/http\"    []Todo\n}\n\n\n\nfunc    \"net/http\" main() {\n    tmpl    \"net/http\" := template.Mus    \"net/http\"t(t)PReiter, r *htt  data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1574}}}
===== received =====
Content-Length: 323

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1574,"diagnostics":[{"range":{"start":{"line":6,"character":4},"end":{"line":6,"character":4}},"severity":1,"source":"syntax","message":"expected declaration, found \"html/template\""}]}}
===== sent =====
Content-Length: 1257

{"method":"textDocument/didChange","jsonrpc":"2.0","params":{"contentChanges":[{"text":"package main\n\nimport (\n\t\"html/template\"\n\t\"net/http\"\n)\n    \"html/template\"\n\n    \"html/template\"\nimport (\n    \"html/template\"\n    \"html/template\"\n    \"html/template\"\n    \"net/http\"\n    \"html/template\"\n)\n    \"html/template\"\n\n    \"html/template\"\ntype Todo struct {\n    \"html/template\"\n    Title string\n    \"html/template\"\n    Done  bool\n}\n    \"html/template\"\n\nfunc main() {\n    \"net/http\"t    \"net/http\"yp    \"net/http\"e TodoPage    \"net/http\"struct {\n    PageTitl    \"net/http\"e string\n    Tod    \"net/http\"os     \"net/http\"    []Todo\n}\n\n\n\nfunc    \"net/http\" main() {\n    tmpl    \"net/http\" := template.Mus    \"net/http\"t(t)PReiter, r *htt  data := TodoPageData{\n            PageTitle: \"My TODO list\",\n            Todos: []Todo{\n                {Title: \"Task 1\", Done: false},\n                {Title: \"Task 2\", Done: true},\n                {Title: \"Task 3\", Done: true},\n            },\n        }\n        tmpl.Execute(w, data)\n    })\n    http.ListenAndServe(\":80\", nil)\n}\n"}],"textDocument":{"uri":"file:///home/shayan/workspace/vimgotest/main.go","version":1575}}}
```

**Comment by shayan-7** [comment]
I should tell you when I took the following steps, the issue with copy and pasting external code disappeared:

1. Uninstall vim
2. Remove the `.vim` directory
3. Install vim
4. Install vim-plug
5. Install go-vim

Now I have to find out which `.vimrc` config would cause this messed up code.
But I have a question for you. What do you mean of `What does :set report when you see this problem?` ?
I mean how could I use `:set` to get more info logs?

**Comment by shayan-7** [comment]
In the case I have this code:

```go
package main

func main() {
    fmt.Print
}
```
And cursor is just after `Print` and I press `Ctrl+n`, the following error message appears at the bottom of screen:

```-- Keyword completion (^N^P) Pattern not found```

![image](https://user-images.githubusercontent.com/39003982/116146743-15a07380-a6f4-11eb-8d35-ba9619f71d72.png)

Is it normal or another issue to report?

**Comment by bhcleek** [comment]
I mean if you execute `:set`, what does it report? I was curious about your vim settings, but that's irrelevant now that you're completely reset.

Should we close this?

**Comment by shayan-7** [comment]
@bhcleek Thank you for every single response. I would close this issue.

**Comment by bheadmaster** [comment]
Sorry for necroposting, but I had this exact problem, found this issue over a search engine, so I thought it might help others:

This problem happened when I (accidentally) inserted spaces instead of tabs inside a .go file.
Replacing all the spaces with tabs fixed it:
```
:set noexpandtab
:%retab!
```

**Comment by wade-welles** [comment]
I have this exact problem; I have files with a lot of comments and I was able to reproduce it in a fresh empty main package by just adding a enumerator and lots of comments and it eventually broke. 

Thank you so much for providing the solution (well haven't tried it yet).

Maybe this is some edge case bug that was re-introduced?

**Comment by wade-welles** [comment]
Can you let me know what portion is likely breaking? lsp? Is there a way I can disable it to get the rest of the functionality?

**Comment by wade-welles** [comment]
Okay I confirmed it is the amount of comments. So I was able to resolve the scrambling issue and re-enabled everything. Then I kept getting the `offset 5189 is past the end of the file 5043` type error. And it was happening on two files with lots of comments, and another file was working but it was small and with little comments.

So I purged the comments slowly and the ratio between the two values eventually flattened out and the error went away. I was able to still have like 10% of my comments. 

So there is some issue in I believe gofmt specifically that can't handle large amounts of comments. I was able to create a new project and produce this issue. 

If you have any reproducible issues or want me to create a new thread regarding this issue let me know; I be happy to help.
